diff --git a/driver/product/kernel/drivers/gpu/arm/midgard/mali_malisw.h b/driver/product/kernel/drivers/gpu/arm/midgard/mali_malisw.h
index 3a4db10..6230843 100644
--- a/driver/product/kernel/drivers/gpu/arm/midgard/mali_malisw.h
+++ b/driver/product/kernel/drivers/gpu/arm/midgard/mali_malisw.h
@@ -52,7 +52,7 @@
  * As a macro it may evaluate its arguments more than once.
  * Refer to MAX macro for more details
  */
-#define MIN(x, y)	((x) < (y) ? (x) : (y))
+//##define MIN(x, y)	((x) < (y) ? (x) : (y))
 
 /**
  * MAX -  Return the greater of two values.
@@ -64,7 +64,7 @@
  * to retrieve the min and max of two values, consider using a conditional swap
  * instead.
  */
-#define MAX(x, y)	((x) < (y) ? (y) : (x))
+//#define MAX(x, y)	((x) < (y) ? (y) : (x))
 
 /**
  * @hideinitializer
