diff --git a/yabause/src/retro_arena/main.cpp b/yabause/src/retro_arena/main.cpp
index d47516ca..5a667e15 100644
--- a/yabause/src/retro_arena/main.cpp
+++ b/yabause/src/retro_arena/main.cpp
@@ -306,7 +306,7 @@ int yabauseinit()
   yinit.rbg_use_compute_shader = pre.getBool( "Use compute shader" , false);
 #endif
 
-  yinit.use_sh2_cache = 1;
+  yinit.use_sh2_cache = 0;
 
   res = YabauseInit(&yinit);
   if( res == -1) {
