diff --git a/yabause/src/sh2_dynarec_devmiyax/DynarecSh2CInterface.cpp b/yabause/src/sh2_dynarec_devmiyax/DynarecSh2CInterface.cpp
index 84e2285a..63bc0c9b 100755
--- a/yabause/src/sh2_dynarec_devmiyax/DynarecSh2CInterface.cpp
+++ b/yabause/src/sh2_dynarec_devmiyax/DynarecSh2CInterface.cpp
@@ -797,6 +797,7 @@ int DebugDelayClock() {
 }
 
 int DebugEachClock() {
+#if 0
   dynaLock();
   #define INSTRUCTION_B(x) ((x & 0x0F00) >> 8)
   #define INSTRUCTION_C(x) ((x & 0x00F0) >> 4)
@@ -811,7 +812,6 @@ int DebugEachClock() {
   
 
   u32 pc = DynarecSh2::CurrentContext->GET_PC();
-#if 0
   u32 pc = DynarecSh2::CurrentContext->GET_PC();
   u16 inst = memGetWord(pc);
   s32 m = INSTRUCTION_C(inst);
@@ -837,7 +837,7 @@ int DebugEachClock() {
     dfp = fopen("instest.txt","w");
   }
   u16 inst = memGetWord(pc);
-
+#endif
 #if 0
   if ((inst & 0xf00f) == 0x300d) {
     s32 m = INSTRUCTION_C(inst);
@@ -852,11 +852,11 @@ int DebugEachClock() {
     );
   }
 #endif
+#if 0
   if (!DynarecSh2::CurrentContext->IsSlave()) {
 
     s32 m = INSTRUCTION_C(inst);
     s32 n = INSTRUCTION_B(inst);
-#if 0
     if ((inst & 0xf00f) == 0x000f) {
       u32 addrn = DynarecSh2::CurrentContext->GetGenRegPtr()[n] - 4;
       u32 addrm = DynarecSh2::CurrentContext->GetGenRegPtr()[m] - 4;
@@ -870,11 +870,11 @@ int DebugEachClock() {
       fflush(dfp);
     }
 #endif
+#if 0
   if (!DynarecSh2::CurrentContext->IsSlave()) {
 
     s32 m = INSTRUCTION_C(inst);
     s32 n = INSTRUCTION_B(inst);
-#if 0
     if ((inst & 0xf00f) == 0x000f) {
       u32 addrn = DynarecSh2::CurrentContext->GetGenRegPtr()[n] - 4;
       u32 addrm = DynarecSh2::CurrentContext->GetGenRegPtr()[m] - 4;
@@ -889,6 +889,7 @@ int DebugEachClock() {
     }
 #endif
 
+#if 0
     // MACW
     if ((inst & 0xf00f) == 0x400f) {
       u32 addrn = DynarecSh2::CurrentContext->GetGenRegPtr()[n] - 2;
@@ -902,6 +903,7 @@ int DebugEachClock() {
       );
       fflush(dfp);
     }
+#endif
 #if 0
     // SHAR
     if ((inst & 0xf0ff) == 0x4021) {
@@ -913,6 +915,7 @@ int DebugEachClock() {
       fflush(dfp);
     }
 #endif
+#if 0
     if ((inst & 0xf00f) == 0x200d) {
       fprintf(dfp, "%08X: xtract R[%d]=%08X,R[%d]=%08X\n",
         pc,
@@ -921,7 +924,8 @@ int DebugEachClock() {
       );
       fflush(dfp);
     }
-#if 1
+#endif
+#if 0
     if (pc == 0x06021588 && 
       DynarecSh2::CurrentContext->GetGenRegPtr()[1] == 0x0000318C &&
       DynarecSh2::CurrentContext->GetGenRegPtr()[2] == 0x318C2108 ) {
@@ -930,9 +934,9 @@ int DebugEachClock() {
     }
 #endif
   }
-#endif
 
 
+#if 0
     // MACW
     if ((inst & 0xf00f) == 0x400f) {
       u32 addrn = DynarecSh2::CurrentContext->GetGenRegPtr()[n] - 2;
@@ -946,7 +950,6 @@ int DebugEachClock() {
       );
       fflush(dfp);
     }
-#if 0
     // SHAR
     if ((inst & 0xf0ff) == 0x4021) {
       fprintf(dfp, "%08X: shar R[%d]=%08X,SR=%08X\n",
@@ -957,6 +960,7 @@ int DebugEachClock() {
       fflush(dfp);
     }
 #endif
+#if 0
     if ((inst & 0xf00f) == 0x200d) {
       fprintf(dfp, "%08X: xtract R[%d]=%08X,R[%d]=%08X\n",
         pc,
@@ -965,7 +969,6 @@ int DebugEachClock() {
       );
       fflush(dfp);
     }
-#if 1
     if (pc == 0x06021588 && 
       DynarecSh2::CurrentContext->GetGenRegPtr()[1] == 0x0000318C &&
       DynarecSh2::CurrentContext->GetGenRegPtr()[2] == 0x318C2108 ) {
@@ -974,7 +977,6 @@ int DebugEachClock() {
     }
 #endif
   }
-#endif
 
 
 #if 0
